File: 0726-Number of Atoms/0726-Number of Atoms.cpp
class Solution {
public:
    string countOfAtoms(string formula) {
        int n = formula.size();
        int i = 0;
        map<string, int> table = helper(formula, i);
        string result;
        for (auto t : table) {
            result += t.first;
            if (t.second != 1) {
                result += to_string(t.second);
            }
        }
        return result;
    }
    
private:
    map<string, int> helper(string& s, int& i) {
        map<string, int> table;
        while (i < s.size() && s[i] != ')') {
            if (s[i] == '(') {
                ++i;
                map<string, int> next = helper(s, i);
                for (auto n : next) {
                    table[n.first] += n.second;
                }
            }
            else {
                int start = i++;
                while (i < s.size() && islower(s[i])) {
                    ++i;
                }
                
                string name = s.substr(start, i - start);
                int num = 0;
                while (i < s.size() && isdigit(s[i])) {
                    num = num * 10 + (s[i] - '0');
                    ++i;
                }
                if (num == 0) {
                    num = 1;
                }
                table[name] += num;
            }
        }

        if (i < s.size()) {
            ++i;
            int num = 0;
            while (i < s.size() && isdigit(s[i])) {
                num = num * 10 + (s[i] - '0');
                ++i;
            }

            if (num != 0) {
                for (auto& t : table) {
                    t.second *= num;
                }
            }
        }        
        
        return table;
    }
};

File: 0701-0800/0726-Number of Atoms/0726-Number of Atoms.py
import collections
class Solution:
    def countOfAtoms(self, formula: str) -> str:
        i = 0
        def dfs(s):
            nonlocal i
            table = collections.Counter()
            while i < len(s) and s[i] != ')':
                if s[i] == '(':
                    i += 1
                    for k, v in dfs(s).items():
                        table[k] += v
                else:
                    start = i
                    i += 1
                    while i < len(s) and s[i].islower():
                        i += 1
                    name = s[start: i]
                    num = 0
                    while i < len(s) and s[i].isdigit():
                        num = num*10 + int(s[i])
                        i += 1
                    if num == 0:
                        num = 1
                    table[name] += num
            
            if i < len(s):
                # s[i] is ')'
                i += 1
                num = 0
                while i < len(s) and s[i].isdigit():
                    num = num*10 + int(s[i])
                    i += 1
                if num != 0:
                    for k in table:
                        table[k] *= num
            return table        
        
        table = dfs(formula)
        return ''.join(k + (str(v) if v > 1 else '') for k, v in sorted(table.items()))

class Solution2:
    def countOfAtoms(self, formula: str) -> str:
        i = 0
        St = [collections.Counter()]
        while i < len(formula):
            if formula[i] == '(':
                i += 1
                St.append(collections.Counter())
            elif formula[i] == ')':
                top = St.pop()
                i += 1
                start = i
                while i < len(formula) and formula[i].isdigit():
                    i += 1
                num = int(formula[start:i] or 1)
                for k, v in top.items():
                    St[-1][k] += v * num
            else:
                start = i
                i += 1
                while i < len(formula) and formula[i].islower():
                    i += 1
                name = formula[start: i]
                start = i
                while i < len(formula) and formula[i].isdigit():
                    i += 1
                num = int(formula[start:i] or 1)
                St[-1][name] += num
                
        return ''.join(k + (str(v) if v > 1 else '') for k, v in sorted(St[-1].items()))
